src/Evolution/Systems/NewtonianEuler/BoundaryCorrections/Hllc.cpp

```diff
@@ -0,0 +1,373 @@
+    const Scalar<DataVector>& normal_dot_velocity_ext,
+    const Scalar<DataVector>& largest_outgoing_char_speed_ext,
+    const Scalar<DataVector>& largest_ingoing_char_speed_ext,
+    const dg::Formulation dg_formulation) const noexcept {
+  // Allocate a temp buffer
+  const size_t vector_size = get(mass_density_int).size();
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>,
+                       ::Tags::TempScalar<2>, ::Tags::TempScalar<3>,
+                       ::Tags::TempScalar<4>>>
+      temps{vector_size};
+
+  // Determine lambda_max and lambda_min from the characteristic speeds from
+  // interior and exterior
+  get(get<::Tags::TempScalar<0>>(temps)) =
+      min(0., get(largest_ingoing_char_speed_int),
+          -get(largest_outgoing_char_speed_ext));
+  const DataVector& lambda_min = get(get<::Tags::TempScalar<0>>(temps));
+  get(get<::Tags::TempScalar<1>>(temps)) =
+      max(0., get(largest_outgoing_char_speed_int),
+          -get(largest_ingoing_char_speed_ext));
+  const DataVector& lambda_max = get(get<::Tags::TempScalar<1>>(temps));
+
+  // Compute lambda_star, the contact wave speed. (cf. Eq 10.37 of Toro2009,
+  // note that here lambda instead of S was used to denote characteristic
+  // speeds)
+  get(get<::Tags::TempScalar<2>>(temps)) =
+      (get(pressure_ext) - get(pressure_int) +
+       get(mass_density_int) * get(normal_dot_velocity_int) *
+           (lambda_min - get(normal_dot_velocity_int)) +
+       get(mass_density_ext) * get(normal_dot_velocity_ext) *
+           (lambda_max + get(normal_dot_velocity_ext))) /
+      (get(mass_density_int) * (lambda_min - get(normal_dot_velocity_int)) -
+       get(mass_density_ext) * (lambda_max + get(normal_dot_velocity_ext)));
+  const DataVector& lambda_star = get(get<::Tags::TempScalar<2>>(temps));
+
+  // Precompute common numerical factors for state variables in star region
+  // (cf. Eq 10.39 of Toro2009)
+  get(get<::Tags::TempScalar<3>>(temps)) =
+      (lambda_min - get(normal_dot_velocity_int)) / (lambda_min - lambda_star);
+  const DataVector& prefactor_int = get(get<::Tags::TempScalar<3>>(temps));
+  get(get<::Tags::TempScalar<4>>(temps)) =
+      (lambda_max + get(normal_dot_velocity_ext)) / (lambda_max - lambda_star);
+  const DataVector& prefactor_ext = get(get<::Tags::TempScalar<4>>(temps));
+
+  for (size_t i = 0; i < vector_size; ++i) {
+    // check if lambda_star falls in the correct range [lambda_min,lambda_max]
+    ASSERT(
+        (lambda_star[i] <= lambda_max[i]) and (lambda_star[i] >= lambda_min[i]),
+        "lambda_star in HLLC boundary correction is not consistent : "
+            << "\n lambda_min  = " << lambda_min[i] << "\n lambda_*    = "
+            << lambda_star[i] << "\n lambda_max  = " << lambda_max[i]);
+
+    if (dg_formulation == dg::Formulation::WeakInertial) {
+      // Compute intermediate flux F_star (cf. Eq 10.71 - 10.73 of Toro2009)
+      if (lambda_star[i] >= 0.0) {
```

Comment:
If I've understood correctly, the reason you only have 2 branches (an `if` and an `else`) instead of the 4 in the documentation is because `lambda_min` is zero in the code when `lambda_min` is positive in the documentation. If that's correct, could you maybe note that either here in the code or add to the documentation at the end something like
```
Note that in the implementation we use
G = {  F_int + lambda_min (U^*_int - U_int)   if ...
          F_ext...
where in this case
lambda_min = min(0, ....)
lambda_max = max(0, ...)
```

Reply:
Yep, that's correct. Once the `lambda_star` passes the ASSERT (line 211) test, we are safe to recover all 4 branches by using the same trick used for HLL solver: using `min(0, ..)` or `max(0,...)` to automatically reduce to the pure upwinding.

I missed this part so I'll add notes to the documentation.

Reply:
But honestly I am not sure it would be better to explicitly split into 4 branches, since for instance when `lambda_min > 0`, we can just simply
```
boundary_correction_# = normal_dot_flux_#_int
```
rather than computing some quantities and then multiplying with zero.

Reply:
I also don't know, but I wouldn't worry about it. I'd say leave the code as-is and add a couple sentences to the documentation since that's easiest.